### DISCUSSION TITLE
Dart 3 is coming around mid-2023

### DISCUSSION THREAD

**Discussion by agordeev** [rrousselGit/freezed#843] [Q&A] [answered]
According to the [announcement](https://medium.com/dartlang/the-road-to-dart-3-afdd580fbefa), the next major release will bring us few features that are covered by freezed:

## Sealed classes

See [spec](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md) 

**Usage:** 
```dart
sealed class Shape {
  double calculateArea();
}

class Square implements Shape {
  final double length;
  Square(this.length);
  double calculateArea() => length * length;
}

class Circle implements Shape {
  final double radius;
  ...
}

double calculateArea(Shape shape) => switch (shape) {
  Square(length: var l) => l * l,
  Circle(radius: var r) => math.pi * r * r
};
```

## Macros (static metaprogramming)

See [spec](https://github.com/dart-lang/language/blob/master/working/macros/feature-specification.md)

This supersedes the code generation.

# Question

@rrousselGit giving all that, how do you see the future of `freezed`?

**Comment by rrousselGit** [comment, marked as answer]
Once Dart has built-in pattern matching+sealed classes, Freezed will probably slowly fade away.

Depending on when metaprogramming is released, I may migrate Freezed to metaprogramming though. As people are unlikely to refactor their Freezed classes into sealed classes in a quick fashion

**Comment by thipokch** [comment]
Since Dart 3 is currently being previewed on Flutter 3.9.0-1.0.pre.2, I'm wondering if anyone have tried running this package on it? I also assume that freezed is still useful for the following:
- override toString, operator ==, hashCode
- implement a copyWith method to clone the object
- handling de/serialization

**Comment by tjarvstrand** [comment]
freezed will continue to provide value, if a bit more limited in scope, in terms of data classes. Kotlin has sealed classes and if/switch expressions and Scala also has pattern matching, but immutable data classes (or case classes in Scala) are still very useful and ubiquitous in both languages

My 2c.
